src/replication.c

```diff
@@ -3390,3 +3418,211 @@ void replicationStartPendingFork(void) {
+    if (host) {
+        client *replica = findReplica(host, port);
+
+        if (replica == NULL) {
+            addReplyError(c,"FAILOVERTO requested destination is not "
+                            "a replica.");
+            return;
+        }
+
+        /* Check if requested replica is online */
+        if (replica->replstate != SLAVE_STATE_ONLINE) {
+            addReplyError(c,"FAILOVERTO requested replica is not online.");
+            return;
+        }
+
+        server.target_replica_host = zstrdup(host);
+        server.target_replica_port = port;
+    }
+
+    mstime_t now = mstime();
+    server.failover_end_time = now + timeout_in_ms;
+    server.force_failover = force_flag;
+    /* Cluster failover pauses for timeout_in_ms*2 so do it here too, we will
+     * unpause early in failoverCron if failover times out. */
+    pauseClients(now + (timeout_in_ms*2),CLIENT_PAUSE_WRITE);
+
+    serverLog(LL_NOTICE,"FAILOVERTO %s:%s (timeout: %ld ms) requested.",
+        (char*) c->argv[1]->ptr, (char*) c->argv[2]->ptr, timeout_in_ms);
+
+    addReply(c,shared.ok);
+}
+
+/* Failover cron function, checks coordinated failover state. */
+void failoverCron(void) {
+    mstime_t now = mstime();
+
+    /* Check if failover already completed */
+    if (server.masterhost && (server.repl_state >= REPL_STATE_TRANSFER)) {
+        serverLog(LL_NOTICE,"FAILOVERTO %s:%d completed successfully.",
+            server.masterhost, server.masterport);
+        goto cron_complete;
+    }
+
+    /* Check if failover operation has timed out */
+    if (server.failover_end_time <= now) {
+        if (server.masterhost) {
+            /* We attempted to failover but it did not complete in time so
+             * become a master again to undo the failover. */
+            serverLog(LL_NOTICE,
+                "FAILOVERTO %s:%d attempted, but sync was unsuccessful.",
+                server.target_replica_host, server.target_replica_port);
+            replicationUnsetMaster();
+            goto cron_complete;
+        } else if (server.force_failover) {
+            /* If timeout has expired force a failover if requested. */
+            replicationSetMaster(server.target_replica_host,
+                server.target_replica_port);
+            serverLog(LL_NOTICE,
+                "FAILOVERTO %s:%d expired, but forcing failover.",
+                server.target_replica_host, server.target_replica_port);
+            /* Extend failover time to allow replication to complete, then reset
+             * the force flag so progress will be checked during the extended
+             * duration. */
+            server.failover_end_time += FAILOVERTO_TIMEOUT;
+            server.force_failover = 0;
+            return;
+        } else {
+            /* Failover did not attempt or succeed. */
+            serverLog(LL_NOTICE,"FAILOVERTO %s:%d was unsuccessful.",
+                server.target_replica_host, server.target_replica_port);
```

Comment:
Yeah, this is the case where the second half of the handshake (waiting to move to state transfer) times out. Since this is getting dropped, this code should be clearer.